src/common/hist_util.cuh

```diff
@@ -197,12 +343,54 @@ HistogramCuts AdapterDeviceSketch(AdapterT* adapter, int num_bins,
   for (auto begin = 0ull; begin < batch.Size();
        begin += sketch_batch_num_elements) {
     size_t end = std::min(batch.Size(), size_t(begin + sketch_batch_num_elements));
-    ProcessBatch(adapter, begin, end, missing, &sketch_container, num_cuts);
+    auto const& batch = adapter->Value();
+    ProcessBatch(batch, adapter->DeviceIdx(), adapter->NumColumns(),
+                 begin, end, missing, &sketch_container, num_cuts);
   }
 
   dense_cuts.Init(&sketch_container.sketches_, num_bins, adapter->NumRows());
   return cuts;
 }
+
+template <typename Batch>
+void AdapterDeviceSketch(Batch batch, int num_bins,
```

Comment:
This is confusing to me, there are two functions with this name that return different things and accept different inputs. Instead of writing a whole new interface for sketching from batches why don't you upgrade the previous interface that accepts an adapter so that it can iterate over multiple batches?

Reply:
Another suggestion would be to differentiate variable names because we have two concepts of a batch (my fault). One comes from the adapter and the other is a subset of the adapter batch. If we used something like "batch_subset" for the second it might be clearer.

Reply:
@RAMitchell 

>  why don't you upgrade the previous interface that accepts an adapter so that it can iterate over multiple batches

We don't have multiple batches in 1 adapter.  In  #5629 we have multiple batches from an external iterator.  Surely we can make an adapter for it.  But then every function have to iterate it independently, resulting into multiple times more iterations.

Reply:
> If we used something like "batch_subset" for the second it might be clearer.

I changed it into `SlidingWindow`.  It's a name used in literature.

Reply:
Also, I want to replace the previous implementation of device DMatrix to the new one in  #5629 .  If this is doable, we won't be needing the old function in next PR.

Reply:
Oh I see, the newer version will not accept an adapter? It's less of a problem if we can get rid of the old interface.

Reply:
>  But then every function have to iterate it independently, resulting into multiple times more iterations.

What functions need to operate on a batch concurrently with sketching before iterating? I thought the sketching would need to iterate over all the data before other things happen?

Reply:
@RAMitchell I can pack multiple things in 1 iteration.  Like getting number of non-zero elements for each batch, while at the same time run sketching on that batch.